build-common/src/org/jetbrains/kotlin/incremental/KotlinClassInfo.kt

```diff
@@ -84,123 +109,148 @@ class KotlinClassInfo constructor(
-    private var classVersion: Int? = null
-
-    override fun visit(version: Int, access: Int, name: String, signature: String?, superName: String?, interfaces: Array<out String>?) {
-        super.visit(version, access, name, signature, superName, interfaces)
-        classVersion = version
+    override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<out String>?): MethodVisitor? {
+        return if (shouldSkipMethod(JvmMemberSignature.Method(name, desc))) {
+            null
+        } else {
+            cv.visitMethod(access, name, desc, signature, exceptions)
+        }
     }
+}
 
-    override fun visitMethod(access: Int, name: String, desc: String, signature: String?, exceptions: Array<out String>?): MethodVisitor? {
-        // Note: Do not filter out private methods here because a *public* inline function may actually have a *private* corresponding JVM
-        // method in the bytecode (see `InlineOnlyKt.isInlineOnlyPrivateInBytecode`).
-        // Just filter the methods based on the given `inlineFunctionsAndAccessors` set.
-        val method = JvmMemberSignature.Method(name, desc)
-        if (method !in inlineFunctionsAndAccessors) return null
+/** Computes the snapshot of a Java class represented by a [ClassNode]. */
+object ClassNodeSnapshotter {
 
+    fun snapshotClass(classNode: ClassNode): Long {
         val classWriter = ClassWriter(0)
+        classNode.accept(classWriter)
+        return classWriter.toByteArray().hashToLong()
+    }
 
-        // The `version` and `name` parameters are important (see KT-38857), the others can be null.
-        classWriter.visit(/* version */ classVersion!!, /* access */ 0, /* name */ "ClassWithOneMethod", null, null, null)
-
-        return object : MethodVisitor(Opcodes.API_VERSION, classWriter.visitMethod(access, name, desc, signature, exceptions)) {
-            override fun visitEnd() {
-                result[method] = classWriter.toByteArray().md5()
-            }
+    fun snapshotClassExcludingMembers(classNode: ClassNode, alsoExcludeKotlinMetaData: Boolean = false): Long {
+        val originalFields = classNode.fields
+        val originalMethods = classNode.methods
+        val originalVisibleAnnotations = classNode.visibleAnnotations
+        classNode.fields = emptyList()
+        classNode.methods = emptyList()
+        if (alsoExcludeKotlinMetaData) {
+            classNode.visibleAnnotations = originalVisibleAnnotations?.filterNot { it.desc == "Lkotlin/Metadata;" }
         }
+        return snapshotClass(classNode).also {
+            classNode.fields = originalFields
+            classNode.methods = originalMethods
+            classNode.visibleAnnotations = originalVisibleAnnotations
+        }
+    }
+
+    fun snapshotField(fieldNode: FieldNode): Long {
+        val classNode = emptyClass()
+        classNode.fields.add(fieldNode)
+        return snapshotClass(classNode)
```

Comment:
This method is used for both snapshotting Java classes and Kotlin constants:
   - When snapshotting Java classes, this is necessary because we need to capture the field's name, type, annotations, etc. in addition to its value.
   - When snapshotting Kotlin constants, you're right that we only need to capture the constant's value (the other info is already captured in the Kotlin metadata/proto). In the next PR ([https://github.com/JetBrains/kotlin/pull/5147](https://github.com/JetBrains/kotlin/pull/5147)), I've captured only the value now. (Btw, before this PR -- i.e. before pull/5127, we also captured only the value.)